### ISSUE TITLE
Investigate possibly broken Docker SIGTERM delivery

### ISSUE THREAD

**Issue by davidopp** [GoogleCloudPlatform/kubernetes#5585]
@rsokolowski and I found the following behavior last week when we were working with someone who was trying to run some stuff on Kubernetes. Note that this appears to be a Docker problem, not a Kubernetes problem, but I'm filing it here for now so we can verify that it's reproducible before filing a bug against Docker.

Start two containers on the same machine, each a shell script that traps SIGTERM and prints something like “I received SIGTERM” and quits. Then do “docker stop” and observe that one container exits via the SIGTERM handler path and the other just gets SIGKILL.

**Comment by thockin** [comment]
Are they both the same container image?  There's a known difference between
containers that use docker's CMD vs ENTRYPOINT statement.

On Tue, Mar 17, 2015 at 9:48 PM, David Oppenheimer <notifications@github.com

**Comment by davidopp** [comment]
Yes, same container image.

I can't remember if the Dockerfile was using CMD or ENTRYPOINT. If Rafal doesn't remember, I can ask the person we were working with.

And to clarify my description of the problem -- "the other just gets SIGKILL" should say "and the other just gets SIGKILL after the grace period has expired"

**Comment by thockin** [comment]
if it is the same container it should behave the same :)

On Tue, Mar 17, 2015 at 9:59 PM, David Oppenheimer <notifications@github.com

**Comment by yujuhong** [comment]
What version of docker and what OS were you using?

I can't reproduce this problem at all. I ran the following command in two terminals.
`docker run -i ubuntu /bin/bash -c "trap 'echo term ; exit' SIGTERM ; echo waiting ;  read"`

Then I `docker stop` both of them, and both processes printed "term" and exit.

**Comment by davidopp** [comment]
Thanks for looking into it, Yu-Ju. I emailed the person who had the problem to tell him we couldn't reproduce it, and here's what he said:

I looked a bit more into it earlier this week and figured out what the problem was. Nothing wrong with docker or kubernetes.

Since the script was doing a sleep 15, SIGTERM was not handled by bash until the sleep subprocess had terminated. Seems like bash is not forwarding signals to subprocesses, and only handles the signals when they exit.

Since the grace period is 10 seconds, we were seeing the nondeterministic behaviour where 2/3 of the time it just got killed. (15-5)/15.

The solution was to run the sleep in the background, keep track of the pid and wait for it with bash builtin wait. This will cause bash to actually handle signals while waiting. So we could then terminate sleep and do the unregistration. It works flawlessly now :)
